ugc/storage.cpp

```diff
@@ -1,33 +1,370 @@
 #include "ugc/storage.hpp"
+#include "ugc/serdes.hpp"
+#include "ugc/serdes_json.hpp"
 
 #include "indexer/feature_decl.hpp"
+#include "indexer/index.hpp"
+
+#include "coding/file_name_utils.hpp"
+#include "coding/file_reader.hpp"
+#include "coding/file_writer.hpp"
+#include "coding/internal/file_data.hpp"
+
+#include "platform/platform.hpp"
+
+#include <algorithm>
+#include <utility>
+
+#include "3party/jansson/myjansson.hpp"
 
 namespace ugc
 {
-Storage::Storage(std::string const & filename)
-  : m_filename(filename)
+using namespace std;
+
+namespace
 {
-  Load();
+string const kIndexFileName = "index.json";
+string const kUGCUpdateFileName = "ugc.update.bin";
+string const kTmpFileExtension = ".tmp";
+
+using Sink = MemWriter<vector<char>>;
+
+string GetUGCFilePath() { return my::JoinPath(GetPlatform().WritableDir(), kUGCUpdateFileName); }
+
+string GetIndexFilePath() { return my::JoinPath(GetPlatform().WritableDir(), kIndexFileName); }
+
+bool GetUGCFileSize(uint64_t & size)
+{
+  return GetPlatform().GetFileSizeByName(kUGCUpdateFileName, size);
+}
+
+void DeserializeUGCIndex(string const & jsonData, vector<Storage::UGCIndex> & res)
+{
+  if (jsonData.empty())
+    return;
+
+  MemReader reader(jsonData.c_str(), jsonData.length());
+  NonOwningReaderSource source(reader);
+  DeserializerJsonV0<NonOwningReaderSource> des(source);
+  des(res);
 }
 
-UGCUpdate const * Storage::GetUGCUpdate(FeatureID const & id) const
+string SerializeUGCIndex(vector<Storage::UGCIndex> const & indexes)
 {
-  auto const it = m_ugc.find(id);
-  if (it != end(m_ugc))
-    return &it->second;
-  return nullptr;
+  if (indexes.empty())
+    return string();
+
+  auto array = my::NewJSONArray();
+  vector<char> data;
+  for (auto const & index : indexes)
+  {
+    data.clear();
+    Sink sink(data);
+    SerializerJson<Sink> ser(sink);
+    ser(index);
+
+    my::Json node(data.data());
+    json_array_append_new(array.get(), node.get());
+  }
+
+  unique_ptr<char, JSONFreeDeleter> buffer(json_dumps(array.get(), JSON_COMPACT | JSON_ENSURE_ASCII));
+  return string(buffer.get());
+}
+
+void SerializeUGCUpdate(json_t * node, UGCUpdate const & update, Storage::UGCIndex const & index)
+{
+  vector<char> data;
+  Sink sink(data);
+  SerializerJson<Sink> ser(sink);
+  ser(update);
+
+  my::Json serializedUgc(data.data());
+  node = serializedUgc.get();
+  auto embeddedNode = my::NewJSONObject();
+  ToJSONObject(*embeddedNode, "data_version", index.m_dataVersion);
+  ToJSONObject(*embeddedNode, "mwm_name", index.m_mwmName);
+  ToJSONObject(*embeddedNode, "feature_id", index.m_featureId);
+  ToJSONObject(*node, "feature", *embeddedNode);
+  embeddedNode.release();
+}
+}  // namespace
+
+Storage::Storage(Index const & index) : m_index(index) { Load(); }
+
+UGCUpdate Storage::GetUGCUpdate(FeatureID const & id) const
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  if (m_UGCIndexes.empty())
+    return {};
+
+  auto const feature = GetOriginalFeature(id);
```

Comment:
Why do you need this? I mean, GetOriginalFeature calls ParseEverything(), but you need to call ParseEverything only in the one case - when you're going to save FeatureType and use it later, when the file reader is dead. So, why, why do you need this GetOriginalFeatureByIndex at all?